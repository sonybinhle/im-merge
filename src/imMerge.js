import {isObject, isEmpty, isMatch} from './utils';

const IM_MERGE_TYPE_KEY = '__IM_MERGE_TYPE__';

const MERGE_TYPES = {
  MERGE: 'MERGE',
  INSERT: 'INSERT',
  INSERT_FIRST: 'INSERT_FIRST',
  INSERT_LAST: 'INSERT_LAST',
  INSERT_BEFORE_MATCH: 'INSERT_BEFORE_MATCH',
  INSERT_AFTER_MATCH: 'INSERT_AFTER_MATCH',
  REMOVE: 'REMOVE',
  REMOVE_FIRST: 'REMOVE_FIST',
  REMOVE_LAST: 'REMOVE_LAST',
  REMOVE_MATCH: 'REMOVE_MATCH'
};

export const mergeType = (data) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.MERGE,
  data,
});

export const insertType = (data, index, flatten) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.INSERT,
  data,
  index,
  flatten
});

export const insertFisrtType = (data, flatten) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.INSERT_FIRST,
  data,
  flatten
});

export const insertLastType = (data, flatten) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.INSERT_LAST,
  data,
  flatten
});

export const insertBeforeMatchType = (data, match, flatten) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.INSERT_BEFORE_MATCH,
  data,
  match,
  flatten
});

export const insertAfterMatchType = (data, match, flatten) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.INSERT_AFTER_MATCH,
  data,
  match,
  flatten
});

export const removeType = (index) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.REMOVE,
  index
});

export const removeFirstType = () => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.REMOVE_FIRST
});

export const removeLastType = () => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.REMOVE_LAST
});

export const removeMatchType = (match) => ({
  [IM_MERGE_TYPE_KEY]: MERGE_TYPES.REMOVE_MATCH,
  match
});

const mergeArrayType = (data, source) => {
  const {
    [IM_MERGE_TYPE_KEY]: type,
    data: sourceData,
    match,
    index = 0,
    flatten = true
  } = source;

  const arraySourceData = Array.isArray(sourceData)
    ? (flatten ? sourceData : [ sourceData ])
    : [ sourceData ];

  switch (type) {
    case MERGE_TYPES.MERGE:
      return data.map((item, index) => imMerge(item, sourceData[index]));
    case MERGE_TYPES.INSERT:
      return [...data.slice(0, index), ...arraySourceData, ...data.slice(index)];
    case MERGE_TYPES.INSERT_FIRST:
      return [...arraySourceData, ...data];
    case MERGE_TYPES.INSERT_LAST:
      return [...data, ...arraySourceData];
    case MERGE_TYPES.INSERT_BEFORE_MATCH:
    case MERGE_TYPES.INSERT_AFTER_MATCH:
      const matchedIndex = data.findIndex(item => isMatch(item, match));
      if (matchedIndex === -1) return data;
      const insertedIndex = type === MERGE_TYPES.INSERT_BEFORE_MATCH ? matchedIndex : matchedIndex + 1;
      return [...data.slice(0, insertedIndex), ...arraySourceData, ...data.slice(insertedIndex)];
    case MERGE_TYPES.REMOVE:
      return [...data.slice(0, index), ...data.slice(index + 1)];
    case MERGE_TYPES.REMOVE_FIRST:
      return data.slice(1);
    case MERGE_TYPES.REMOVE_LAST:
      return data.slice(0, data.length - 1);
    case MERGE_TYPES.REMOVE_MATCH:
      return data.filter(item => !isMatch(item, match));
    default:
      return data;
  }
};

const imMerge = (data, source) => {
  if (Array.isArray(data) && isObject(source) && source[IM_MERGE_TYPE_KEY]) {
    return mergeArrayType(data, source);
  }
  if (Array.isArray(data) && Array.isArray(source)) {
    if (!source.length) return data;

    return [...data, ...source];
  }
  if (isObject(data) && isObject(source) && !(Array.isArray(data) || Array.isArray(source))) {
    if (isEmpty(source)) return data;

    return Object.keys(source).reduce((acc, key) => {
      acc[key] = imMerge(acc[key], source[key]);

      return acc;
    }, { ...data });
  }

  return source;
};

export default imMerge;
