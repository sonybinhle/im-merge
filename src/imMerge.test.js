import imMerge, {
  insertType,
  insertFisrtType,
  insertLastType,
  removeType,
  removeFirstType,
  removeLastType
} from './imMerge';

describe('imMerge', () => {
  it('when merge data with source --> then return new deeply merged object', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: { f: '3' } } };
    const source = { a: { b: { c: 2, dd: 3 }, ee: { f: '4' } } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
  });

  it('when merge data with empty source --> then return original object', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: { f: '3' } } };
    const source = { };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toBe(data);
  });

  it('when merge data with child empty source --> then return original child object', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: { f: '3' } } };
    const source = { a: { e: {} } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj.a.e).toBe(data.a.e);
    expect(mergedObj.a).not.toBe(data.a);
  });

  it('when merge data with some non-collision keys --> then non-collision objects will return original references', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: { f: '3' } } };
    const source = { a: { b: { c: 2 }, e: {} }, d: {} };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj.a.e).toBe(data.a.e);
    expect(mergedObj.a).not.toBe(data.a);
    expect(mergedObj.a.b.c).toBe(source.a.b.c);
    expect(mergedObj.a.b.d).toBe(data.a.b.d);
    expect(mergedObj.d).toBe(source.d);
    expect(mergedObj.d).not.toBe(data.d);
  });

  it('when merge data source array --> then default behavior is appended', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: [{ x: 1 }, { x: 2 }] } };
    const source = { a: { b: { c: 2 }, e: [{ x: 3 }] } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.a.e[0]).toBe(data.a.e[0]);
    expect(mergedObj.a.e[1]).toBe(data.a.e[1]);
    expect(mergedObj.a.e[2]).toBe(source.a.e[0]);
  });

  it('when source is a empty array --> then return original data', () => {
    // given
    const data = { a: { b: { c: 1, d: 2 }, e: [{ x: 1 }, { x: 2 }] } };
    const source = { a: { b: { c: 2 }, e: [] } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj.a.e).toBe(data.a.e);
    expect(mergedObj.a.e.length).toBe(2);
  });
});

describe('imMerge-array', () => {
  it('when merge array with insertType source array --> then default index is 0 and return new array with inserted object', () => {
    // given
    const data = { a: { b: {}, t: [{ x: 1 }, { x: 2 }] } };
    const insertedData = { x: 3 };
    const source = { a: { b: { c: 2 }, t: insertType(insertedData) } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.a.t[0]).toBe(insertedData);
    expect(mergedObj.a.t[1]).toBe(data.a.t[0]);
    expect(mergedObj.a.t[2]).toBe(data.a.t[1]);
  });

  it('when merge array with insertType and index --> then return new array with inserted object', () => {
    // given
    const data = { a: { b: {}, t: [{ x: 1 }, { x: 2 }] } };
    const insertedData = { x: 3 };
    const source = { a: { b: { c: 2 }, t: insertType(insertedData, 1) } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.a.t[0]).toBe(data.a.t[0]);
    expect(mergedObj.a.t[1]).toBe(insertedData);
    expect(mergedObj.a.t[2]).toBe(data.a.t[1]);
  });

  it('when merge array with insertType, index, default flatten --> then return new array with inserted objects', () => {
    // given
    const data = { a: { b: {}, t: [{ x: 1 }, { x: 2 }] } };
    const insertedData = [ { x: 3 }, { x: 4 } ];
    const source = { a: { b: { c: 2 }, t: insertType(insertedData, 1) } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.a.t[0]).toBe(data.a.t[0]);
    expect(mergedObj.a.t[1]).toBe(insertedData[0]);
    expect(mergedObj.a.t[2]).toBe(insertedData[1]);
    expect(mergedObj.a.t[3]).toBe(data.a.t[1]);
  });

  it('when merge array with insertType, index, flatten=false --> then return new array with inserted objects', () => {
    // given
    const data = { a: { b: {}, t: [{ x: 1 }, { x: 2 }] } };
    const insertedData = [ { x: 3 }, { x: 4 } ];
    const source = { a: { b: { c: 2 }, t: insertType(insertedData, 1, false) } };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.a.t[0]).toBe(data.a.t[0]);
    expect(mergedObj.a.t[1]).toBe(insertedData);
    expect(mergedObj.a.t[2]).toBe(data.a.t[1]);
  });

  it('when merge array with insertFirstType --> then return new array with inserted object', () => {
    // given
    const data = { t: [{ x: 1 }, { x: 2 }, { x: 3 }] };
    const insertedData = { x: 4 };
    const source = { t: insertFisrtType(insertedData) };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.t[0]).toBe(insertedData);
    expect(mergedObj.t[1]).toBe(data.t[0]);
  });

  it('when merge array with insertLastType --> then return new array with inserted object', () => {
    // given
    const data = { t: [{ x: 1 }, { x: 2 }, { x: 3 }] };
    const insertedData = { x: 4 };
    const source = { t: insertLastType(insertedData) };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.t[2]).toBe(data.t[2]);
    expect(mergedObj.t[3]).toBe(insertedData);
  });

  it('when merge array with removeType --> then return new array with object removed', () => {
    // given
    const data = { t: [{ x: 1 }, { x: 2 }, { x: 3 }] };
    const source = { t: removeType(1) };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.t[0]).toBe(data.t[0]);
    expect(mergedObj.t[1]).toBe(data.t[2]);
  });

  it('when merge array with removeFirstType --> then return new array with object removed', () => {
    // given
    const data = { t: [{ x: 1 }, { x: 2 }, { x: 3 }] };
    const source = { t: removeFirstType() };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.t[0]).toBe(data.t[1]);
    expect(mergedObj.t[1]).toBe(data.t[2]);
  });

  it('when merge array with removeLastType --> then return new array with object removed', () => {
    // given
    const data = { t: [{ x: 1 }, { x: 2 }, { x: 3 }] };
    const source = { t: removeLastType() };

    // when
    const mergedObj = imMerge(data, source);

    // then
    expect(mergedObj).toMatchSnapshot();
    expect(mergedObj.t[0]).toBe(data.t[0]);
    expect(mergedObj.t[1]).toBe(data.t[1]);
    expect(mergedObj.t[3]).toBe(undefined);
  });
});
