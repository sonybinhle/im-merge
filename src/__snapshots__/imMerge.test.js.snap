// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imMerge when merge data source array --> then default behavior is appended 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
      "d": 2,
    },
    "e": Array [
      Object {
        "x": 1,
      },
      Object {
        "x": 2,
      },
      Object {
        "x": 3,
      },
    ],
  },
}
`;

exports[`imMerge when merge data with source --> then return new deeply merged object 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
      "d": 2,
      "dd": 3,
    },
    "e": Object {
      "f": "3",
    },
    "ee": Object {
      "f": "4",
    },
  },
}
`;

exports[`imMerge-array when merge array with insertFirstType --> then return new array with inserted object 1`] = `
Object {
  "t": Array [
    Object {
      "x": 4,
    },
    Object {
      "x": 1,
    },
    Object {
      "x": 2,
    },
    Object {
      "x": 3,
    },
  ],
}
`;

exports[`imMerge-array when merge array with insertLastType --> then return new array with inserted object 1`] = `
Object {
  "t": Array [
    Object {
      "x": 1,
    },
    Object {
      "x": 2,
    },
    Object {
      "x": 3,
    },
    Object {
      "x": 4,
    },
  ],
}
`;

exports[`imMerge-array when merge array with insertType and index --> then return new array with inserted object 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
    },
    "t": Array [
      Object {
        "x": 1,
      },
      Object {
        "x": 3,
      },
      Object {
        "x": 2,
      },
    ],
  },
}
`;

exports[`imMerge-array when merge array with insertType source array --> then default index is 0 and return new array with inserted object 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
    },
    "t": Array [
      Object {
        "x": 3,
      },
      Object {
        "x": 1,
      },
      Object {
        "x": 2,
      },
    ],
  },
}
`;

exports[`imMerge-array when merge array with insertType, index, default flatten --> then return new array with inserted objects 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
    },
    "t": Array [
      Object {
        "x": 1,
      },
      Object {
        "x": 3,
      },
      Object {
        "x": 4,
      },
      Object {
        "x": 2,
      },
    ],
  },
}
`;

exports[`imMerge-array when merge array with insertType, index, flatten=false --> then return new array with inserted objects 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 2,
    },
    "t": Array [
      Object {
        "x": 1,
      },
      Array [
        Object {
          "x": 3,
        },
        Object {
          "x": 4,
        },
      ],
      Object {
        "x": 2,
      },
    ],
  },
}
`;

exports[`imMerge-array when merge array with removeFirstType --> then return new array with object removed 1`] = `
Object {
  "t": Array [
    Object {
      "x": 2,
    },
    Object {
      "x": 3,
    },
  ],
}
`;

exports[`imMerge-array when merge array with removeLastType --> then return new array with object removed 1`] = `
Object {
  "t": Array [
    Object {
      "x": 1,
    },
    Object {
      "x": 2,
    },
  ],
}
`;

exports[`imMerge-array when merge array with removeType --> then return new array with object removed 1`] = `
Object {
  "t": Array [
    Object {
      "x": 1,
    },
    Object {
      "x": 3,
    },
  ],
}
`;
